### DISCUSSION TITLE
How to use preset-env in ES2020

### DISCUSSION THREAD

**Discussion by josemigallas** [babel/babel#15609] [Q&A] [answered]
According to the docs, [babel-plugin-proposal-nullish-coalescing-operator](https://babeljs.io/docs/babel-plugin-proposal-nullish-coalescing-operator.html) is included in babel/preset-env _in ES2020_. But what does this mean? How can it be enabled?

I added `browserslist` to my package.json and set `browserslistConfigFile: true` in babel.config.js, with the following support:
```
"browserslist": [
  "chrome >= 107",
  "firefox >= 105",
  "edge >= 107"
],
```
And still I keep getting errors like:
```
ERROR in ./app/javascript/packs/invoice_payment.ts 86:40
Module parse failed: Unexpected token (86:40)
File was processed with these loaders:
 * ./node_modules/babel-loader/lib/index.js
You may need an additional loader to handle the result of these loaders.
|   card.on('change', event => {
|     payButton.disabled = !event.complete;
>     cardError.textContent = event.error?.message ?? '';
|   });
|   form.addEventListener('submit', function (event) {
```

**Comment by JLHwung** [comment, marked as answer]
```
"browserslist": [
  "chrome >= 107",
  "firefox >= 105",
  "edge >= 107"
]
```
Your target already supports `?.`, that's why `preset-env` does not enable the transform for you. However, in your case, the webpack 4 does not support it. Please upgrade to webpack 5.

If you are stuck at webpack 4, you can use `include` to always enable `?.` transform, though the output code will be slightly bigger than what it supposed to be.

```json
["@babel/preset-env", { "include": [ "proposal-optional-chaining" ] }]
```

**Reply by josemigallas** [reply]
I see, and good point I should upgrade webpack ASAP.

My question though would still be: is there an explicit way of setting ES2020 as target for Babel?

**Reply by nicolo-ribaudo** [reply]
No, because ECMAScript version don't correspond in practice with real-world engines or tools. We used to have "yearly presets", but [they have been removed](https://babeljs.io/blog/2017/12/27/nearing-the-7.0-release.html#deprecated-yearly-presets-eg-babel-preset-es20xx). If you really want to enable a specific ECMAScript version, you could manually list the corresponding plugins.

**Reply by JLHwung** [reply]
> is there an explicit way of setting ES2020 as target for Babel

Nope. We are not planning to support it at this moment as browser versions are more friendly to end users than a technical term like ES2020.

**Reply by dietergeerts** [reply]
@JLHwung , I have a situation where we need to transpile for the QJSEngine used by Qt Installer framework, which doesn't support all features. For example, the object spread operator is missing. For this situation, I can't use any of the currently supported targets, because it's a custom engine, however, the engine does specify a certain ES version it supports. Thus having babel supporting setting a specific ES version as target would be very useful. Unless there's a way to have some sort of translation from an ES version to some target babel does support?

**Reply by JLHwung** [reply]
As for QuickJS, there is a PR (https://github.com/compat-table/compat-table/pull/1665) to add it to the compat-table, the upstream data source of babel/compat-data. Once compat-table supports it, Babel might support qjs as a supported target.

Before that happens, you can check out https://babel.dev/docs/plugins-list and manually add plugins you need according to the ES versions.
